simulation/conveyor/model.py

```diff
@@ -0,0 +1,325 @@
+            if speed is None:
+                speed = self._speed
+            objs = shift(self._object_positions, after * speed)
+        else:
+            objs = self._object_positions
+
+        res = search_pos(objs, pos, preference=preference)
+        if res is not None:
+            (oid, o_pos), idx = res
+            if not_exact and o_pos == pos:
+                if preference == 'prev':
+                    idx -= 1
+                elif preference == 'next':
+                    idx += 1
+                else:
+                    raise Exception('please dont use nearest with not exact')
+                if idx < 0 or idx > len(objs):
+                    return None
+                oid, o_pos = objs[idx]
+            return self._objects[oid], o_pos
+        return None
+
+    def putObject(self, obj_id: int, obj: Any, pos: float, soft_collide=True, return_nearest=False):
+        assert obj_id not in self._objects, "Clashing object ID!"
+        pos = round(pos, POS_ROUND_DIGITS)
+
+        nearest = None
+        if len(self._objects) > 0:
+            (n_obj_id, n_pos), n_idx = search_pos(self._object_positions, pos)
+            if n_pos == pos:
+                if soft_collide:
+                    print(f'{self._model_id}: TRUE COLLISION: #{obj_id} and #{n_obj_id} on {pos}')
+                    i = n_idx
+                    p_pos = pos
+                    while i < len(self._object_positions) and self._object_positions[i][1] >= p_pos:
+                        p_pos = round(p_pos + SOFT_COLLIDE_SHIFT, POS_ROUND_DIGITS)
+                        self._object_positions[i] = (self._object_positions[i][0], p_pos)
+                        i += 1
+                else:
+                    raise CollisionException((obj, self._objects[n_obj_id], pos, self._model_id))
+            elif n_pos < pos:
+                n_idx += 1
+            nearest = (n_obj_id, n_pos)
+        else:
+            n_idx = 0
+
+        self._objects[obj_id] = obj
+        self._object_positions.insert(n_idx, (obj_id, pos))
+        self._stateTransfer('change')
+
+        if return_nearest:
+            return nearest
+
+    def removeObject(self, obj_id: int):
+        pos_idx = None
+        for (i, (oid, pos)) in enumerate(self._object_positions):
+            if oid == obj_id:
+                pos_idx = i
+                break
+
+        self._object_positions.pop(pos_idx)
+        obj = self._objects.pop(obj_id)
+        self._stateTransfer('change')
+        return obj
+
+    def shift(self, d):
+        self._object_positions = shift(self._object_positions, d)
+
+    def skipTime(self, time: float, clean_ends=True):
+        if time == 0:
+            return 0
+
+        self._stateTransfer('change')
+        d = time * self._speed
+        if len(self._objects) == 0:
+            return d
+
+        self.shift(d)
+
+        if clean_ends:
+            while len(self._object_positions) > 0 and self._object_positions[0][1] < 0:
```

Comment:
Maybe it is better to use some structure for single `object_position`? I am slightly afraid of  `self._object_positions[0][1]`